<a href="https://colab.research.google.com/github/MelisaCHasan/MAT461/blob/main/Module%20A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Problem #1 Converting binary to decimal

In [15]:
def my_bin_2_dec(b):
    d = 0
    element = b
    power = len(element) - 1
    for i in b:
      x = 2**power * i
      power -= 1
      d += x
    return d


In [16]:
my_bin_2_dec([1, 1, 1])

7

In [17]:
my_bin_2_dec([1, 0, 1, 0, 1, 0, 1])

85

In [19]:
my_bin_2_dec([1]*25)

33554431

Problem #2 Converting decimal to binary


In [46]:
def my_dec_2_bin(d):
    b = []
    if d >= 1:
      while d > 0:
        b.append(d % 2)
        d = d // 2
      b.reverse()
    else:
      b.append(0)
    return b


[1, 0, 1, 1, 1]

In [47]:
my_dec_2_bin(0)

[0]

In [48]:
my_dec_2_bin(23)

[1, 0, 1, 1, 1]

In [49]:
my_dec_2_bin(2097)

[1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1]

Problem #3 Floating Point to Decimal

In [68]:
def my_ieee_2_dec(ieee):
    sign_bit = int(ieee[0])
    exponent_bits = ieee[1:12]
    fraction_bits = ieee[12:]
    exponent = int(exponent_bits,2)
    bias = 1023
    exp = exponent - bias
    fraction = 1.0
    for i in range(52):
      #print(fraction)
      fraction += float(fraction_bits[i]) * 2**(-i-1)
    d = ((-1)**sign_bit) * (2**exp) * fraction
    print(d)

    return d

In [65]:
ieee = '1100000001001000000000000000000000000000000000000000000000000000'
my_ieee_2_dec(ieee)

-48.0


-48.0

In [66]:
ieee = '0100000000001011001100110011001100110011001100110011001100110011'
my_ieee_2_dec(ieee)

3.4


3.4

The code above is correct but for the second problem I am experiencing round off errors. If you removethe comment from print(fraction) you can see that the fraction does not go beyond a certain number of bits before it just rounds to give 1.7 as the fraction which gives the rounded number 3.4 instead of the expected 3.39999999999999991118215802999.